src/yubikey.rs

```diff
@@ -128,1223 +90,261 @@ pub struct Version {
+            .data(&[0x7c, 0x02, 0x80, 0x00])
+            .transmit(&txn, 261)?;
 
-            free(self.pin as (*mut c_void));
-            self.pin = ptr::null_mut();
+        if !challenge.is_success() || challenge.buffer().len() < 12 {
+            return Err(Error::AuthenticationError);
         }
 
-        if !pin.is_null() && len > 0 {
-            self.pin = malloc(len + 1) as (*mut u8);
+        // send a response to the cards challenge and a challenge of our own.
+        let response = mgm_key.decrypt(challenge.buffer()[4..12].try_into().unwrap());
 
-            if self.pin.is_null() {
-                return Err(Error::MemoryError);
-            }
+        let mut data = [0u8; 22];
+        data[0] = 0x7c;
+        data[1] = 20; // 2 + 8 + 2 +8
+        data[2] = 0x80;
+        data[3] = 8;
+        data[4..12].copy_from_slice(&response);
+        data[12] = 0x81;
+        data[13] = 8;
 
-            memcpy(self.pin as (*mut c_void), pin as (*const c_void), len);
-            *self.pin.add(len) = 0u8;
+        if getrandom(&mut data[14..22]).is_err() {
+            error!("failed getting randomness for authentication");
+            return Err(Error::RandomnessError);
         }
 
-        Ok(())
-    }
+        let mut challenge = [0u8; 8];
+        challenge.copy_from_slice(&data[14..22]);
 
-    /// Verify device PIN
-    ///
-    /// Returns the number of tries remaining both on success and on a wrong PIN.
-    pub unsafe fn ykpiv_verify(&mut self, pin: *const c_char) -> Result<i32, Error> {
-        self.ykpiv_verify_select(pin, if !pin.is_null() { strlen(pin) } else { 0 }, false)
-    }
+        let authentication = APDU::new(YKPIV_INS_AUTHENTICATE)
+            .params(YKPIV_ALGO_3DES, YKPIV_KEY_CARDMGM)
+            .data(&data)
+            .transmit(&txn, 261)?;
 
-    /// Verify device PIN
-    ///
-    /// Returns the number of tries remaining both on success and on a wrong PIN.
-    pub(crate) unsafe fn _verify(
-        &mut self,
-        pin: *const c_char,
-        pin_len: usize,
-    ) -> Result<i32, Error> {
-        let mut data = [0u8; 261];
-        let mut recv_len = data.len() as u32;
-        let mut sw: i32 = 0;
-
-        if pin_len > CB_PIN_MAX {
-            return Err(Error::SizeError);
+        if !authentication.is_success() {
+            return Err(Error::AuthenticationError);
         }
 
-        let mut apdu = APDU::new(YKPIV_INS_VERIFY);
-        apdu.params(0x00, 0x80);
-
-        if !pin.is_null() {
-            let mut data = [0xFF; CB_PIN_MAX];
-
-            memcpy(
-                data.as_mut_ptr() as *mut c_void,
-                pin as *const c_void,
-                pin_len,
-            );
+        // compare the response from the card with our challenge
+        let response = mgm_key.encrypt(&challenge);
 
-            apdu.data(data);
+        use subtle::ConstantTimeEq;
+        if response.ct_eq(&authentication.buffer()[4..12]).unwrap_u8() != 1 {
```

Comment:
[Use the `From` implementation instead](https://docs.rs/subtle/2.2.1/subtle/struct.Choice.html#note):
```suggestion
        if !response.ct_eq(&authentication.buffer()[4..12]).into() {
```

Reply:
This doesn't compile:

```
error[E0282]: type annotations needed
   --> src/yubikey.rs:264:61
    |
264 |         if !response.ct_eq(&authentication.buffer()[4..12]).into() {
    |                                                             ^^^^ cannot infer type for `T`
    |
    = note: type must be known at this point
```